File: problems/evaluate_division/evaluate_division.cc
#include "evaluate_division.h"

#include <cstddef>
#include <string>
#include <unordered_map>
#include <vector>

using std::string;
using std::unordered_map;
using std::vector;

namespace {
std::unordered_map<std::string, std::string> parent;
std::unordered_map<std::string, double> weight;

// NOLINTBEGIN(misc-no-recursion)
string find(const string& node, unordered_map<string, string>& parent,
            unordered_map<string, double>& weight) {
    if (parent[node] != node) {
        const string originalParent = parent[node];
        parent[node] = find(originalParent, parent, weight);
        weight[node] *= weight[originalParent];
    }
    return parent[node];
}
// NOLINTEND(misc-no-recursion)
}  // namespace

vector<double> evaluateDivision(const vector<vector<string>>& equations,
                                const vector<double>& values,
                                const vector<vector<string>>& queries) {
    // Clear previous state
    parent.clear();
    weight.clear();

    const size_t numEquations = equations.size();

    for (const auto& equation : equations) {
        parent[equation[0]] = equation[0];
        parent[equation[1]] = equation[1];
        weight[equation[0]] = 1.0;
        weight[equation[1]] = 1.0;
    }

    for (size_t i = 0; i < numEquations; ++i) {
        const vector<string>& equation = equations[i];
        const string& numerator = equation[0];
        const string& denominator = equation[1];

        const string parentNumerator = find(numerator, parent, weight);
        const string parentDenominator = find(denominator, parent, weight);

        if (parentNumerator == parentDenominator) {
            continue;
        }

        parent[parentNumerator] = parentDenominator;
        weight[parentNumerator] = values[i] * weight[denominator] / weight[numerator];
    }

    const size_t numQueries = queries.size();
    vector<double> results(numQueries);

    for (size_t i = 0; i < numQueries; ++i) {
        const vector<string>& query = queries[i];
        const string& numerator = query[0];
        const string& denominator = query[1];

        if (!parent.contains(numerator) || !parent.contains(denominator) ||
            find(numerator, parent, weight) != find(denominator, parent, weight)) {
            results[i] = -1.0;
        } else {
            results[i] = weight[numerator] / weight[denominator];
        }
    }

    return results;
}

File: problems/evaluate_division/evaluate_division.py
from typing import List
from collections import defaultdict


def evaluateDivision(
    equations: List[List[str]], values: List[float], queries: List[List[str]]
) -> List[float]:
    def find(node: str) -> str:
        if parent[node] != node:
            original_parent = parent[node]
            parent[node] = find(original_parent)
            weight[node] *= weight[original_parent]
        return parent[node]

    weight = defaultdict(lambda: 1.0)
    parent = defaultdict(str)

    for numerator, denominator in equations:
        parent[numerator] = numerator
        parent[denominator] = denominator

    for (numerator, denominator), value in zip(equations, values):
        root_numerator = find(numerator)
        root_denominator = find(denominator)

        if root_numerator == root_denominator:
            continue

        parent[root_numerator] = root_denominator
        weight[root_numerator] = value * weight[denominator] / weight[numerator]

    results = []
    for numerator, denominator in queries:
        if (
            numerator not in parent
            or denominator not in parent
            or find(numerator) != find(denominator)
        ):
            results.append(-1.0)
        else:
            results.append(weight[numerator] / weight[denominator])

    return results
